addons/payment/controllers/portal.py

```diff
@@ -74,7 +74,14 @@ def payment_pay(
 
         # Raise an HTTP 404 if a partner is provided with an invalid access token
         if partner_id:
-            if not payment_utils.check_access_token(access_token, partner_id, amount, currency_id):
+            token_verified = (payment_utils.check_access_token(access_token, partner_id, amount,
+                                                               currency_id, kwargs['confirm_order'])
+                              if 'confirm_order' in kwargs else
+                              payment_utils.check_access_token(access_token, partner_id, amount,
+                                                               currency_id)
+                              )
+
+            if not token_verified:
```

Comment:
This should be done in the sale part. You can create a new token minus the `confirm_order` once you've verified it in the sale controller. Doing this, no need to change anything in payment